### PR TITLE
Per tenant access to tenant and cluster objects

### PR DESCRIPTION
**Pull request by corvus-ch**
Implements #111

## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the documentation.
- [x] Update the ./CHANGELOG.md.
- [x] Link this PR to related issues.

### CODE REVIEW

File: pkg/apis/syn/v1alpha1/tenant_types.go [hunk 1]
```diff
@@ -122,3 +125,100 @@ func (t *Tenant) ApplyTemplate(template *TenantTemplate) error {
...
+
+// UpdateRole updates Role so it meets the needs as an object to grant access to the tenant.
+// The function will return true, if the role got altered.
+func (t *Tenant) UpdateRole(role *rbacv1.Role) bool {
+	return roleUpdateResourceNames(role, t.Name)
+}
+
+// CreateRoleBinding returns a new RoleBinding object.
+// Its intention is to link subjects to the role created with `Tenant.CreateRole()`.
+func (t *Tenant) CreateRoleBinding() *rbacv1.RoleBinding {
```
Comment (ccremer) [inline pkg/apis/syn/v1alpha1/tenant_types.go:162]: also here
```suggestion
func (t *Tenant) NewRoleBinding() *rbacv1.RoleBinding {
```

**Review by ccremer — CHANGES_REQUESTED** [review]

File: pkg/pipeline/cluster_steps.go [hunk 2]
```diff
@@ -144,3 +144,32 @@ func applyClusterTemplateFromTenant(obj PipelineObject, data *ExecutionContext)
...
+
+func clusterUpdateRole(obj PipelineObject, data *ExecutionContext) ExecutionResult {
+	cluster, ok := obj.(*synv1alpha1.Cluster)
+	if !ok {
+		return ExecutionResult{Err: fmt.Errorf("object is not a cluster")}
+	}
+
+	name := types.NamespacedName{Name: cluster.Spec.TenantRef.Name, Namespace: cluster.Namespace}
+	role := &rbacv1.Role{}
+	if err := data.Client.Get(context.TODO(), name, role); err != nil {
```
[the remark this answers is not in the document]
Comment (corvus-ch) [reply pkg/pipeline/cluster_steps.go:156]: No, there is no context available.

File: pkg/apis/syn/v1alpha1/tenant_types_test.go [hunk 3]
```diff
@@ -82,3 +84,211 @@ func TestTenant_ApplyTemplate(t *testing.T) {
...
+
+	assert.Equal(t, []string{synv1alpha1.SchemeGroupVersion.Group}, role.Rules[0].APIGroups)
+	assert.Len(t, role.Rules[0].Resources, 2)
+	assert.Contains(t, role.Rules[0].Resources, "tenants")
+	assert.Contains(t, role.Rules[0].Resources, "clusters")
+	assert.Len(t, role.Rules[0].Verbs, 1)
+	assert.Contains(t, role.Rules[0].Verbs, "get")
+}
+
+func TestTenant_CreateRoleBinding(t *testing.T) {
```
Comment (ccremer) [inline pkg/apis/syn/v1alpha1/tenant_types_test.go:128]: it probably got forgotten
```suggestion
func TestTenant_NewRoleBinding(t *testing.T) {
```

**Review by ccremer — APPROVED** [review]

**Comment by corvus-ch** [comment]
I have now also added an owner reference. @ccremer please have a close look at that one. I am not certain this is the way to do it.

File: pkg/apis/syn/v1alpha1/tenant_types.go [hunk 4]
```diff
@@ -122,3 +125,121 @@ func (t *Tenant) ApplyTemplate(template *TenantTemplate) error {
...
+}
+
+func setOwnerReference(obj metav1.Object, owner *Tenant) {
+	obj.SetOwnerReferences([]metav1.OwnerReference{
+		{
+			APIVersion: SchemeGroupVersion.Identifier(),
+			Kind: owner.Kind,
+			Name: owner.Name,
+		},
+	})
```
Comment (ccremer) [inline pkg/apis/syn/v1alpha1/tenant_types.go:234]: owner.Kind is actually empty if the object was created with "New...".
But there's a utility function
```suggestion
	controllerutil.SetOwnerReference(owner, obj, scheme)
```
We may need more and more to have a common scheme variable somewhere, like https://github.com/vshn/k8up/blob/master/main.go#L33

File: pkg/apis/syn/v1alpha1/tenant_types_test.go [hunk 5]
```diff
@@ -82,3 +84,221 @@ func TestTenant_ApplyTemplate(t *testing.T) {
...
+	assert.Equal(t, value, v)
+}
+
+func assertOwnerReference(t *testing.T, obj v1.Object, owner *synv1alpha1.Tenant) {
+	gvk := synv1alpha1.SchemeGroupVersion.WithKind(owner.Kind)
+	refs := obj.GetOwnerReferences()
+	require.Len(t, refs, 1)
+	assert.Equal(t, gvk.GroupVersion().String(), refs[0].APIVersion)
+	assert.Equal(t, gvk.Kind, refs[0].Kind)
+	assert.Equal(t, owner.GetName(), refs[0].Name)
```
Comment (ccremer) [inline pkg/apis/syn/v1alpha1/tenant_types_test.go:303]: When using the utility function from controller-runtime, asserting all properties may be overkill
```suggestion
	refs := obj.GetOwnerReferences()
	require.Len(t, refs, 1)
	assert.Equal(t, owner.GetName(), refs[0].Name)
```

File: pkg/controller/tenant/tenant_rbac.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	synv1alpha1 "github.com/projectsyn/lieutenant-operator/pkg/apis/syn/v1alpha1"
+	roleUtil "github.com/projectsyn/lieutenant-operator/pkg/role"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// NewServiceAccount returns a new ServiceAccount object.
+// The ServiceAccount has its meta data set in a way, that it is clear this ServiceAccount is related to the tenant.
```
Comment (ccremer) [inline pkg/controller/tenant/tenant_rbac.go:14]:
```suggestion
// The ServiceAccount has its metadata set in a way, that it is clear this ServiceAccount is related to the tenant.
```

File: pkg/controller/tenant/tenant_rbac_test.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	assertOwnerReference(t, sa, tenant)
+}
+
+func TestReconcileTenant_NewRole(t *testing.T) {
+	tenant := &synv1alpha1.Tenant{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-tenant",
+			Namespace: "my-namespace",
+		},
+	}
```
Comment (ccremer) [inline pkg/controller/tenant/tenant_rbac_test.go:61]: This piece seems to be used more often. Maybe add a helper `newTenant()` in the test suite ?

Comment (corvus-ch) [reply pkg/controller/tenant/tenant_rbac_test.go:61]: I have had that tought too. But then, sometimes more than just the object meta is needed. To me it felt like over apstracting, so I left it as is.

**Review by ccremer — APPROVED** [review]

**Pull request merged**